integration_tests/httpcache/main.go

```diff
@@ -0,0 +1,1143 @@
+		calledAfterSend = true
+
+		r.SetBodyTransform(func(r io.ReadCloser) io.ReadCloser {
+			return errReader{}
+		})
+
+		r.SetUncacheable(true)
+		return nil
+	}
+
+	r.CacheOptions.AfterSend = afterSend
+
+	resp, err := r.Send(ctx, backend)
+	if err != nil {
+		return fmt.Errorf("r.Send() err=%v, want %v", err, nil)
+	}
+
+	if !calledAfterSend {
+		return errors.New("calledAfterSend=false")
+	}
+
+	if _, err := io.ReadAll(resp.Body); !errors.Is(err, errReaderErr) {
+		return fmt.Errorf("resp.Body.Read() err=%v, want %v", err, errReaderErr)
+	}
+
+	return nil
+}
+
+func testAfterSendMutations(ctx context.Context) error {
+	r := getTestReq("", nil, nil)
+
+	r.Header.Set("fastly-debug", "1")
+
+	var calledAfterSend bool
+	r.CacheOptions.AfterSend = func(r *fsthttp.CandidateResponse) error {
+		calledAfterSend = true
+		r.SetStatus(201)
+		r.SetHeader("x-custom", "test")
+		r.SetTTL(4000)
+		r.SetStaleWhileRevalidate(400)
+		r.SetSensitive(true)
+		r.SetSurrogateKeys("key1,key2")
+		r.SetVary("accept,user-agent")
+		r.SetCacheable()
+		return nil
+	}
+
+	resp, err := r.Send(ctx, backend)
+	if err != nil {
+		return err
+	}
+
+	if !calledAfterSend {
+		return errors.New("calledAfterSend=false")
+	}
+
+	if got, want := resp.StatusCode, 201; got != want {
+		return fmt.Errorf("response.StatusCode=%v want %v", got, want)
+	}
+
+	if got, want := resp.Header.Get("x-custom"), "test"; got != want {
+		return fmt.Errorf("r.Header.Get('x-custom'): got %q, want %q", got, want)
+	}
+
+	// TODO(dgryski): javascript allow this, but rust doesn't
+	/*
+		if got, ok := resp.StaleWhileRevalidate(); !ok || got != 400 {
+			return fmt.Errorf("response.stale()=%v,%v want %v, %v", got, ok, 400, true)
+		}
+		if got, ok := resp.TTL(); !ok || got != 4000 {
+			return fmt.Errorf("response.ttl()=%v, %v, want %v, %v", got, ok, 4000, true)
+		}
+	*/
+
+	if got, want := resp.Vary(), "accept,user-agent"; got != want {
+		return fmt.Errorf("response.Vary()=%q, want %q", got, want)
+	}
+
+	if got, want := resp.SurrogateKeys(), "key1,key2"; got != want {
+		return fmt.Errorf("response.SurrogateKeys()=%q, want %q", got, want)
+	}
+
+	return nil
+}
+
+func testAfterSendHeaderRemove(ctx context.Context) error {
+	r := getTestReq("", nil, nil)
+
+	var calledAfterSend bool
+	r.CacheOptions.AfterSend = func(r *fsthttp.CandidateResponse) error {
+		calledAfterSend = true
+		r.SetHeader("custom", "custom-header")
+		r.DelHeader("access-control-allow-origin")
+		r.DelHeader("access-control-allow-credentials")
+		r.SetUncacheable(true)
+		return nil
+	}
+
+	resp, err := r.Send(ctx, backend)
+	if err != nil {
+		return err
+	}
+
+	if !calledAfterSend {
+		return errors.New("calledAfterSend=false")
+	}
+
+	if got, want := resp.Header.Get("custom"), "custom-header"; got != want {
+		return fmt.Errorf("r.Header.Get('custom'): got %v, want %v", got, want)
+	}
+
+	if got, want := resp.Header.Get("access-control-allow-origin"), ""; got != want {
```

Comment:
I wasn't confident that this header would always be present, but that's probably just me.